### PR TITLE
Reserve allocation should be displayed when erroring due to lack of memory on startup

### LINKED ISSUE #11168: [BUG] reserve allocation should be displayed when erroring due to lack of memory on startup
Our QA found the following error message as they were debugging something else, and rightly called out it was confusing:

```
24/02/06 04:02:51 ERROR RapidsExecutorPlugin: Exception in the executor plugin, shutting down! 
java.lang.IllegalArgumentException: The pool allocation of -498.875 MB (calculated from
spark.rapids.memory.gpu.allocFraction (=1.0) and 141.125 MB free memory) was less than the minimum allocation of
20252.015625 (calculated from spark.rapids.memory.gpu.minAllocFraction (=0.25) and 81008.0625 MB total memory)
```

The reason for the negative number is we are missing our reserve amount from the message. There are other things to clean up here like that there are no units in the minimum allocation umber. Also these should all be MiB not MB.

We could say:
```
The pool allocation of -498.875 MiB (gpu.free: 141.125 MiB, spark.rapids.memory.gpu.allocFraction: 1.0,
spark.rapids.memory.gpu.reserve: 640 MiB => (gpu.free - reserve) * allocFraction = -498.875 MiB) was less than the minimum
allocation of 20252.016 MiB (gpu.total: 81008.063 MiB, spark.rapids.memory.gpu.minAllocFraction: 0.25 => gpu.total *
minAllocFraction = 20252.016 MiB). Please ensure that the GPU has enough free memory, or adjust configuration accordingly.
```

Added some line breaks in the messages above to make it easier to read.

### PR DESCRIPTION
**Pull request by kuhushukla**
Fixes #11168 
I think using 
```
s"${RapidsConf.RMM_ALLOC_RESERVE}: ${reserveAmount} => " +
```
is the right decision here but I would like for someone to correct or confirm. 
I don't have a test for this. Can look into it if preferred for the review.

### CODE REVIEW

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDeviceManager.scala [hunk 1]
```diff
@@ -240,19 +240,27 @@ object GpuDeviceManager extends Logging {
         (conf.rmmAllocFraction * (info.free - reserveAmount)).toLong)
       if (poolAllocation < minAllocation) {
         throw new IllegalArgumentException(s"The pool allocation of " +
-            s"${toMB(poolAllocation)} MB (calculated from ${RapidsConf.RMM_ALLOC_FRACTION} " +
-            s"(=${conf.rmmAllocFraction}) and ${toMB(info.free)} MB free memory) was less than " +
-            s"the minimum allocation of ${toMB(minAllocation)} (calculated from " +
-            s"${RapidsConf.RMM_ALLOC_MIN_FRACTION} (=${conf.rmmAllocMinFraction}) " +
-            s"and ${toMB(info.total)} MB total memory)")
+            s"${toMB(poolAllocation)} MiB (gpu.free: ${toMB(info.free)}," +
```
Comment (abellina) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDeviceManager.scala:243]: should we rename `toMB` to `toMiB` (nit pick)

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDeviceManager.scala [hunk 2]
```diff
@@ -240,19 +240,27 @@ object GpuDeviceManager extends Logging {
...
+            s"${RapidsConf.RMM_ALLOC_FRACTION}: (=${conf.rmmAllocFraction}," +
+            s"${RapidsConf.RMM_ALLOC_RESERVE}: ${reserveAmount} => " +
+            s"(gpu.free - reserve) * allocFraction = ${toMB(poolAllocation)})" +
+            s"was less than allocation of ${toMB(minAllocation)} MiB (gpu.total: " +
```
Comment (abellina) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDeviceManager.scala:247]: I think we can bring this message into two: the part before "was less than" and the part after. That way we can reuse it in the second case (it is almost verbatim except for this text, as far as I can tell).

Comment (kuhushukla) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDeviceManager.scala:247]: agree

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDeviceManager.scala (hunk 1 above)
Comment (kuhushukla) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDeviceManager.scala:243]: I noticed but left it out to be conservative. Will update

**Comment by kuhushukla** [comment]
@abellina I think this is pretty close, whenever you have a chance to take a look. Thank u!

**Comment by kuhushukla** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDeviceManager.scala [hunk 3]
```diff
@@ -238,33 +238,39 @@ object GpuDeviceManager extends Logging {
...
-            s"${toMB(poolAllocation)} MB (calculated from ${RapidsConf.RMM_ALLOC_FRACTION} " +
-            s"(=${conf.rmmAllocFraction}) and ${toMB(info.free)} MB free memory) was less than " +
-            s"the minimum allocation of ${toMB(minAllocation)} (calculated from " +
-            s"${RapidsConf.RMM_ALLOC_MIN_FRACTION} (=${conf.rmmAllocMinFraction}) " +
-            s"and ${toMB(info.total)} MB total memory)")
+        throw new IllegalArgumentException(errorPhrase +
+            s"less than allocation of ${toMiB(minAllocation)} MiB (gpu.total: " +
+            s"${toMiB(info.total)} MiB, ${RapidsConf.RMM_ALLOC_MIN_FRACTION}: " +
+            s"${conf.rmmAllocMinFraction} => gpu.total *" +
+            s"minAllocFraction = ${toMiB(minAllocation)} MiB). Please ensure that the GPU has" +
```
Comment (abellina) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDeviceManager.scala:251]:
```suggestion
            s"minAllocFraction = ${toMiB(minAllocation)} MiB). Please ensure that the GPU has " +
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDeviceManager.scala [hunk 4]
```diff
@@ -238,33 +238,39 @@ object GpuDeviceManager extends Logging {
...
-            s"${toMB(poolAllocation)} MB (calculated from ${RapidsConf.RMM_ALLOC_FRACTION} " +
-            s"(=${conf.rmmAllocFraction}) and ${toMB(info.free)} MB free memory) was more than " +
-            s"the maximum allocation of ${toMB(maxAllocation)} (calculated from " +
-            s"${RapidsConf.RMM_ALLOC_MAX_FRACTION} (=${conf.rmmAllocMaxFraction}) " +
-            s"and ${toMB(info.total)} MB total memory)")
+        throw new IllegalArgumentException(errorPhrase +
+            s"more than allocation of ${toMiB(maxAllocation)} MiB (gpu.total: " +
+            s"${toMiB(info.total)} MiB, ${RapidsConf.RMM_ALLOC_MAX_FRACTION}: " +
+            s"${conf.rmmAllocMaxFraction} => gpu.total *" +
+            s"maxAllocFraction = ${toMiB(maxAllocation)} MiB). Please ensure that pool allocation" +
```
Comment (abellina) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDeviceManager.scala:259]:
```suggestion
            s"maxAllocFraction = ${toMiB(maxAllocation)} MiB). Please ensure that pool allocation " +
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDeviceManager.scala [hunk 5]
```diff
@@ -238,33 +238,39 @@ object GpuDeviceManager extends Logging {
...
+            s"larger than maximum pool size (${toMiB(maxAllocation)} MB). Check the settings for " +
             s"${RapidsConf.RMM_ALLOC_MAX_FRACTION} (=${conf.rmmAllocFraction}) and " +
             s"${RapidsConf.RMM_ALLOC_RESERVE} (=$reserveAmount)")
       }
       val adjustedMaxAllocation = truncateToAlignment(maxAllocation - reserveAmount)
       if (poolAllocation > adjustedMaxAllocation) {
-        logWarning(s"RMM pool allocation (${toMB(poolAllocation)} MB) does not leave enough free " +
-            s"memory for reserve memory (${toMB(reserveAmount)} MB), lowering the pool size to " +
-            s"${toMB(adjustedMaxAllocation)} MB to accommodate the requested reserve amount.")
+        logWarning(s"RMM pool allocation (${toMiB(poolAllocation)} MB) does not leave enough" +
```
Comment (abellina) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDeviceManager.scala:270]:
```suggestion
        logWarning(s"RMM pool allocation (${toMiB(poolAllocation)} MB) does not leave enough " +
```

**Review by abellina — COMMENTED** [review]
Some missed spaces at the end of strings, but otherwise LGTM

**Review by revans2 — DISMISSED** [review]

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDeviceManager.scala (hunk 3 above)
Comment (kuhushukla) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDeviceManager.scala]: addressed

**Review by abellina — APPROVED** [review]

**Comment by kuhushukla** [comment]
build

**Pull request merged**